### ISSUE TITLE
The type of SectionBase.data as ReadonlyArray

### ISSUE THREAD

**Issue by GregoryEsberci** [facebook/react-native#42337]
### Description

I don't know if this is really a bug, but I couldn't find a better template for it.

I'm using a TypeScript library that returns data as a list, but the data type extends only the `Iterable` type, causing a type error because the `data` type in `SectionList` should be a `ReadonlyArray`.

https://github.com/facebook/react-native/blob/6c8dfc89566d27c36f9ca3828bcf3a63dfcb6942/packages/react-native/Libraries/Lists/SectionList.d.ts#L32

Shouldn't this type be `ArrayLike` just like it is in the `data` of `FlatList`?

https://github.com/facebook/react-native/blob/6c8dfc89566d27c36f9ca3828bcf3a63dfcb6942/packages/react-native/Libraries/Lists/FlatList.d.ts#L47

### Steps to reproduce

Attempting to use an ArrayLike in a SectionList.

### React Native Version

0.73.2

### Affected Platforms

Other (please specify)

### Output of `npx react-native info`

```text
System:
  OS: macOS 13.6
  CPU: (8) arm64 Apple M1
  Memory: 50.94 MB / 16.00 GB
  Shell:
    version: "5.9"
    path: /bin/zsh
Binaries:
  Node:
    version: 18.18.2
    path: ~/.nvm/versions/node/v18.18.2/bin/node
  Yarn:
    version: 1.22.21
    path: ~/.nvm/versions/node/v18.18.2/bin/yarn
  npm:
    version: 9.8.1
    path: ~/.nvm/versions/node/v18.18.2/bin/npm
  Watchman: Not Found
Managers:
  CocoaPods:
    version: 1.14.3
    path: /Users/gregory/.rvm/gems/ruby-2.7.6/bin/pod
SDKs:
  iOS SDK:
    Platforms:
      - DriverKit 22.4
      - iOS 16.4
      - macOS 13.3
      - tvOS 16.4
      - watchOS 9.4
  Android SDK: Not Found
IDEs:
  Android Studio: 2022.2 AI-222.4459.24.2221.10121639
  Xcode:
    version: 14.3.1/14E300c
    path: /usr/bin/xcodebuild
Languages:
  Java:
    version: 11.0.15
    path: /usr/bin/javac
  Ruby:
    version: 2.7.6
    path: /Users/gregory/.rvm/rubies/ruby-2.7.6/bin/ruby
npmPackages:
  "@react-native-community/cli": Not Found
  react:
    installed: 18.2.0
    wanted: 18.2.0
  react-native:
    installed: 0.73.2
    wanted: 0.73.2
  react-native-macos: Not Found
npmGlobalPackages:
  "*react-native*": Not Found
Android:
  hermesEnabled: true
  newArchEnabled: false
iOS:
  hermesEnabled: true
  newArchEnabled: false
```

### Stacktrace or Logs

```text
No overload matches this call.
  Overload 1 of 2, '(props: SectionListProps<Task, { data: ResultsType<Task>; }> | Readonly<SectionListProps<Task, { data: ResultsType<Task>; }>>): SectionList<...>', gave the following error.
    Type 'ResultsType<Task>' is not assignable to type 'readonly Task[] & ResultsType<Task>'.
      Type 'Results<Task>' is not assignable to type 'readonly Task[]'.
        The types of '[Symbol.unscopables].length' are incompatible between these types.
          Type 'number' is not assignable to type 'boolean | undefined'.
  Overload 2 of 2, '(props: SectionListProps<Task, { data: ResultsType<Task>; }>, context: any): SectionList<Task, { data: ResultsType<Task>; }>', gave the following error.
    Type 'ResultsType<Task>' is not assignable to type 'readonly Task[] & ResultsType<Task>'.ts(2769)
```

### Reproducer

https://github.com/GregoryEsberci/rn-flat-list-data-type

**Comment by NickGerleman** [comment]
I looked at the code for this one.

VirtualizedList originally also accepted `ReadonlyArray` but we ended up making it more permissive after we found real usage which relied on it accepting more than just arrays. This was also slightly breaking because the list provided callbacks with the same time, which needed to be made more generic.

`VirtualizedSectionList` is typed to accept arrays, and these arrays are also given to callbacks, to render the sections.

The internals could be changed to operate on ArrayLike instead, but this would be breaking to existing functions which accept sections to render, and assume a full array is present.

TLDR: the current typings are correct, but arguably a larger (but breaking) change could be made to make the data more permissive.

If your data is small enough to copy the `Iterable` to an array, that would be the easiest way around this.
